onnxruntime/core/session/inference_session.cc

```diff
@@ -2181,6 +2191,50 @@ const profiling::Profiler& InferenceSession::GetProfiling() const {
   return session_profiler_;
 }
 
+#if !defined(ORT_MINIMAL_BUILD)
+std::vector<TuningResults> InferenceSession::GetTuningResults() const {
+  std::vector<TuningResults> ret;
+  for (const auto& provider : execution_providers_) {
+    const auto* tuning_ctx = provider->GetTuningContext();
+    if (tuning_ctx != nullptr) {
+      ret.emplace_back(tuning_ctx->SaveTuningResults());
+    }
+  }
+  return ret;
+}
+
+Status InferenceSession::SetTuningResults(const std::vector<TuningResults>& trs, bool error_on_invalid) {
+  std::string msg;
+
+  for (size_t i = 0; i < trs.size(); i++) {
+    const auto& tr = trs[i];
+    auto* provider = execution_providers_.Get(tr.ep);
+    if (provider == nullptr) {
+      msg = MakeString("Cannot find execution provider ", tr.ep);
+      LOGS(*session_logger_, WARNING) << msg;
+      ORT_RETURN_IF(error_on_invalid, msg);
```

Comment:
Who sets `error_on_invalid`? More importantly, should we return if we don't find the EP? For example, we could first run a model on CUDA, save the tuning results for CUDA, and then run it on ROCM and save the tuning results for ROCM. Now the model metadata contains tuning results for both CUDA and ROCM, which is okay. But the code here will reject it, which is not the right behavior. Thoughts?

Reply:
The TuningContext is bind to the EP now. 

```cpp
struct TuningResults { // TR for short
  std::string ep;
  std::unordered_map<std::string, std::string> validators;
  std::unordered_map<std::string, KernelMap> results;
};
```
we use the `ep` to identify which EP to load the TR. However, `ep` may not be correct, because maybe human error of text editing. `validators` may have new keys in new version of ORT. `error_on_invalid=true` is to error out in this case.

However, in the desired scenario, we may dump
```bash
[TR_cuda_v_1.xx] # in TR_cuda_v_1.xx.json
[TR_rocm_v_1.xx] # in TR_cuda_v_1.yy.json

[TR_cuda_v_1.yy]
[TR_rocm_v_1.yy]
```

we then can merge them as 
```bash
[TR_cuda_v_1.xx, TR_rocm_v_1.xx, TR_cuda_v_1.yy, TR_rocm_v_1.yy] # TR_merged.json
```

Our users can blindly load it and benefit from our offline tuning. or we may embed the TRs to onnx file.

So we want to be forgiving =). user should not touch this option in general. And this is why it defaults to `False` in python api.

Reply:
That makes sense. Thanks for the detailed explanation.